### PR TITLE
HTML structure readme section

### PR DESCRIPTION
**Pull request by Voycawojka**
## Description
Relates to #51: How are HTML pages processed?

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+
+```html
+<body>
+  <header>
+    Weather Service
+  </header>
+  
+  <!-- data-knotx-id="auto-generated-id" -->
+  <script data-knotx-debug="log" data-knotx-id="auto-generated-id" type="application/json">{ debug-data-here }</script>
+
```
Comment (tomaszmichalak) [inline README.md]: Do we need this empty line here?

File: README.md [hunk 2]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+  <span>
+    -2°C in Reykjavík
+  </span>
+
```
Comment (tomaszmichalak) [inline README.md]: Do we need this empty line here?

File: README.md [hunk 3]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+      {{fetch-weather-info._result.temperatue}} in {{fetch-weather-info._result.location}}
+    </span>
+  </knotx:snippet>
+</body>
 ```
-<!-- data-knotx-id="FRAGMENT_IDENTIFIER" -->
-  FRAGMENT_BODY
-<!-- data-knotx-id="FRAGMENT_IDENTIFIER" -->
+
+Could result in:
```
Comment (tomaszmichalak) [inline README.md]:
```suggestion
could result in:
```

File: README.md [hunk 4]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+<body>
+  <header>
+    Weather Service
+  </header>
+  
+  <!-- data-knotx-id="auto-generated-id" -->
+  <script data-knotx-debug="log" data-knotx-id="auto-generated-id" type="application/json">{ debug-data-here }</script>
+
+  <span>
+    -2°C in Reykjavík
```
Comment (tomaszmichalak) [inline README.md]: > Reykjavík

nice place ;)

File: README.md [hunk 5]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+  </span>
+
+  <!-- data-knotx-id="auto-generated-id" -->
+</body>
 ```
 
-Fragment details are stored in `<script>` tag:
+Note that the browser displays the page as usual, the injected information have no effect on the presentation.
```
Comment (tomaszmichalak) [inline README.md]:
```suggestion
Note that the browser displays the page as usual, the injected information does not affect the presentation.
```

File: README.md [hunk 6]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+There are two parts of the debug information:
+- The data that was part of the Knot.x snippet is surrounded by special comments with auto-generated IDs. This is mainly used by the tool to provide highligthing functionality.
+- The snippet content is preceded by a script tag with a JSON inside. This is where all the important data is stored. It primarily contains information about the processing of the fragment nodes: which nodes were triggered, what transition they activated, how much time they took to process etc.
```
Comment (tomaszmichalak) [inline README.md]:
```suggestion
- The script tag with a JSON inside precedes the fragment content.  All important data is stored here. It mainly contains information about the processing of the fragment nodes: which nodes were triggered, what transition they activated, how much time they took etc.
```

File: README.md (hunk 1 above)
Comment (Voycawojka) [reply README.md]: I thought it's more readable, but I can change it if you insist

File: README.md (hunk 2 above)
Comment (Voycawojka) [reply README.md]: Same logic as with the previous empty line

File: README.md [hunk 7]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+    },
+    // ...
+  ]
+}
+```
+
+It is then ready to be displayed in the form of a graph (specifically a tree unless there are composite nodes in the fragment).
+
+Internaly the parser consists of two phases:
+- flattening - The fragment's graph is normally a tree. However, composite nodes reference subtasks which are another tree each. This phase creates a new graph structure where all the nodes are part of this graph (there are no sub-graphs)
```
Comment (tomaszmichalak) [inline README.md]:
```suggestion
- flattening - The fragment's graph is normally a tree (an undirected graph in which any two vertices are connected by exactly one path). However, composite nodes reference subtasks which are another tree each. This phase creates a new graph structure where all the nodes are part of this graph (there are no sub-graphs).
```

File: README.md [hunk 8]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+      // ...
+      "on": {
+        "_subtask_end": "composite-node_virtual_end" // note this is only an ID (!)
+      }
+    }
+  }
+}
+```
+
+Important thing to note is that, while all subtasks end with a transition to the `composite-node_virtual_end` node, only one of them (the deepest) contains an actual object in the transition.
```
Comment (tomaszmichalak) [inline README.md]:
```suggestion
An important thing to note is that, while all subtasks end with a transition to the `composite-node_virtual_end` node, only one of them (the deepest) contains an actual object in the transition.
```

File: README.md [hunk 9]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+All other subtasks end with a transition into a string. It's termed `a reference` in the code and it's an ID of the actual node.
+It is like that in order to avoid duplication, especially during the second phase.
```
Comment (tomaszmichalak) [inline README.md]:
```suggestion
It is like that to avoid duplication, especially during the second phase.
```

Comment (tomaszmichalak) [inline README.md]: Can we please add some short example here?

File: README.md [hunk 10]
```diff
@@ -26,26 +26,237 @@ the instructions below:
...
+
+#### Timeline parser
+
+The `timeline/declarationHelper.js` parses a given fragment's debug JSON into a form understandable by (Vis.js Timeline)[https://visjs.github.io/vis-timeline/docs/timeline/], a library for displaying Gantt charts.
```
Comment (tomaszmichalak) [inline README.md]:
```suggestion
The `timeline/declarationHelper.js` parses a given fragment's debug JSON (from the fragment's script tag) into a form understandable by (Vis.js Timeline)[https://visjs.github.io/vis-timeline/docs/timeline/], a library for displaying Gantt charts.
```

**Pull request merged**